### ISSUE TITLE
is nonblocking example blocking?

### ISSUE THREAD

**Issue by Jack079** [knolleary/pubsubclient#738]
Hi,

I am at the point that I want to add MQTT to my project.
If WiFi connection or connection to broker fails, the rest of the program has to work.
I used the non blocking example, changed is a bit for esp32 and wifi.
If WiFi is on and the broker is available is works perfect, But when connection is not available,
it is blocking. 
I made the sketch as small as possible and added some serialprints to see what happens.
After WiFi is connected, it go's 4 times true main loop and after that the sketch is blocked.

Can you help me out?

`
#include <WiFi.h>
#include <PubSubClient.h>

const char* ssid = "****";
const char* password = "****";
const char* mqtt_server = "****";

WiFiClient espClient;
PubSubClient client(espClient);
long lastReconnectAttempt = 0;

boolean reconnect() {
  if (client.connect("arduinoClient")) {
  // Once connected, publish an announcement...
  client.publish("outTopic/two","hello world");
  }
  return client.connected();
  }

void setup()
{
  Serial.begin(9600);
  client.setServer(mqtt_server, 1883);
  connect_WiFi(); //on a other tab
}

void loop()
{
 Serial.println("begin loop");

  connect_WiFi();  //on a other tab

  Serial.println("mqttConnect");

  if (!client.connected()) {
    long now = millis();
    if (now - lastReconnectAttempt > 5000) {
      lastReconnectAttempt = now;
      // Attempt to reconnect
      if (reconnect()) {
        lastReconnectAttempt = 0;
      }
    }
  } else {
    // Client connected
    client.loop();
  }
}
`

**Comment by systembolaget** [comment]
Better use a [state machine approach](https://esp32.com/viewtopic.php?f=19&t=3851#p17925) to WiFi and MQTT connection/re-connection to keep all other code running.

**Comment by pqkhanh561** [comment]
No, in function connect there is a while loop to wait client available until timeout
